File: unreal/AIArenaSDK/Source/AIArenaSDK/Private/Telemetry/TelemetryCollector.cpp
#include "Telemetry/TelemetryCollector.h"
#include "Core/AIArenaApiClient.h"
#include "Core/AIArenaConfig.h"
#include "Dom/JsonObject.h"
#include "Dom/JsonValue.h"
#include "Serialization/JsonWriter.h"
#include "Serialization/JsonSerializer.h"
#include "TimerManager.h"
#include "Engine/World.h"
#include "Engine/GameInstance.h"
#include "HAL/PlatformTime.h"
#include "Misc/Guid.h"

void UAIArenaTelemetryCollector::Initialize(FSubsystemCollectionBase& Collection)
{
    Super::Initialize(Collection);
}

void UAIArenaTelemetryCollector::Deinitialize()
{
    if (GetGameInstance())
    {
        GetGameInstance()->GetTimerManager().ClearTimer(FlushTimerHandle);
    }
    // Best-effort flush on shutdown
    Flush();
    Super::Deinitialize();
}

void UAIArenaTelemetryCollector::StartSession(const FString& SessionId,
                                               const FString& AgentId)
{
    ActiveSessionId = SessionId;
    ActiveAgentId   = AgentId;
    SequenceNumber  = 0;
    RetryCount      = 0;
    EventBuffer.Reset();
    SchedulePeriodicFlush();
    UE_LOG(LogTemp, Log,
           TEXT("[TelemetryCollector] Session started: %s"), *SessionId);
}

void UAIArenaTelemetryCollector::EndSession()
{
    GetGameInstance()->GetTimerManager().ClearTimer(FlushTimerHandle);
    Flush();
    ActiveSessionId.Empty();
    ActiveAgentId.Empty();
}

void UAIArenaTelemetryCollector::SchedulePeriodicFlush()
{
    const UAIArenaConfig* Cfg = UAIArenaConfig::Get();
    GetGameInstance()->GetTimerManager().SetTimer(
        FlushTimerHandle,
        FTimerDelegate::CreateUObject(this, &UAIArenaTelemetryCollector::Flush),
        Cfg->TelemetryFlushIntervalSeconds, /*bLoop=*/true);
}

// ── Event helpers ──────────────────────────────────────────────────────────────

static FString Vec3Json(FVector V)
{
    return FString::Printf(TEXT("{\"x\":%.2f,\"y\":%.2f,\"z\":%.2f}"), V.X, V.Y, V.Z);
}

void UAIArenaTelemetryCollector::RecordCombatAction(const FString& ActionType,
                                                     const FString& TargetId,
                                                     FVector Position,
                                                     bool bSuccess,
                                                     float DamageDealt,
                                                     float LatencyMs)
{
    FString Payload = FString::Printf(
        TEXT("{\"actionType\":\"%s\",\"targetId\":\"%s\","
             "\"position\":%s,\"success\":%s,"
             "\"damageDealt\":%.2f,\"latencyMs\":%.1f}"),
        *ActionType, *TargetId, *Vec3Json(Position),
        bSuccess ? TEXT("true") : TEXT("false"),
        DamageDealt, LatencyMs);
    Enqueue(TEXT("COMBAT_ACTION"), Payload);
}

void UAIArenaTelemetryCollector::RecordPositionUpdate(FVector Position,
                                                       FVector Velocity,
                                                       float Rotation)
{
    FString Payload = FString::Printf(
        TEXT("{\"position\":%s,\"velocity\":%s,\"rotation\":%.2f}"),
        *Vec3Json(Position), *Vec3Json(Velocity), Rotation);
    Enqueue(TEXT("POSITION_UPDATE"), Payload);
}

void UAIArenaTelemetryCollector::RecordAbilityUse(const FString& AbilityId,
                                                   const FString& AbilityName,
                                                   const FString& TargetId,
                                                   float CooldownMs)
{
    FString Payload = FString::Printf(
        TEXT("{\"abilityId\":\"%s\",\"abilityName\":\"%s\","
             "\"targetId\":\"%s\",\"cooldownMs\":%.1f}"),
        *AbilityId, *AbilityName, *TargetId, CooldownMs);
    Enqueue(TEXT("ABILITY_USE"), Payload);
}

void UAIArenaTelemetryCollector::RecordHealthChange(float PreviousHp,
                                                     float CurrentHp,
                                                     float MaxHp,
                                                     const FString& ChangeReason,
                                                     const FString& SourceId)
{
    FString Payload = FString::Printf(
        TEXT("{\"previousHp\":%.1f,\"currentHp\":%.1f,\"maxHp\":%.1f,"
             "\"changeReason\":\"%s\",\"sourceId\":\"%s\"}"),
        PreviousHp, CurrentHp, MaxHp, *ChangeReason, *SourceId);
    Enqueue(TEXT("HEALTH_CHANGE"), Payload);
}

void UAIArenaTelemetryCollector::RecordKill(const FString& VictimId,
                                             const FString& WeaponUsed,
                                             FVector Position,
                                             int32 KillStreak)
{
    FString Payload = FString::Printf(
        TEXT("{\"victimId\":\"%s\",\"weaponUsed\":\"%s\","
             "\"position\":%s,\"killStreak\":%d}"),
        *VictimId, *WeaponUsed, *Vec3Json(Position), KillStreak);
    Enqueue(TEXT("KILL"), Payload);
}

// ── Internal ──────────────────────────────────────────────────────────────────

void UAIArenaTelemetryCollector::Enqueue(const FString& EventType,
                                          const FString& PayloadJson)
{
    const UAIArenaConfig* Cfg = UAIArenaConfig::Get();
    if (!Cfg->bEnableTelemetry || ActiveSessionId.IsEmpty()) return;

    FAIArenaTelemetryEvent Evt;
    Evt.EventId        = FGuid::NewGuid().ToString();
    Evt.SessionId      = ActiveSessionId;
    Evt.AgentId        = ActiveAgentId;
    Evt.EventType      = EventType;
    Evt.TimestampMs    = (int64)(FPlatformTime::Seconds() * 1000.0);
    Evt.SequenceNumber = SequenceNumber++;
    Evt.PayloadJson    = PayloadJson;

    EventBuffer.Add(Evt);

    if (EventBuffer.Num() >= Cfg->TelemetryBatchSize)
    {
        Flush();
    }
}

void UAIArenaTelemetryCollector::FlushImmediate()
{
    Flush();
}

void UAIArenaTelemetryCollector::Flush()
{
    if (EventBuffer.IsEmpty() || ActiveSessionId.IsEmpty()) return;

    // Snapshot and clear the buffer
    TArray<FAIArenaTelemetryEvent> Batch = MoveTemp(EventBuffer);
    EventBuffer.Reset();

    // Build JSON array of events
    FString EventsJson = TEXT("[");
    for (int32 i = 0; i < Batch.Num(); i++)
    {
        const FAIArenaTelemetryEvent& E = Batch[i];
        EventsJson += FString::Printf(
            TEXT("{\"eventId\":\"%s\",\"sessionId\":\"%s\",\"agentId\":\"%s\","
                 "\"eventType\":\"%s\",\"timestamp\":%lld,\"seq\":%d,"
                 "\"payload\":%s}"),
            *E.EventId, *E.SessionId, *E.AgentId,
            *E.EventType, E.TimestampMs, E.SequenceNumber,
            *E.PayloadJson);
        if (i < Batch.Num() - 1) EventsJson += TEXT(",");
    }
    EventsJson += TEXT("]");

    FString BatchId = FString::Printf(
        TEXT("%s-%lld"), *ActiveSessionId,
        (int64)(FPlatformTime::Seconds() * 1000.0));

    FString Body = FString::Printf(
        TEXT("{\"batchId\":\"%s\",\"sessionId\":\"%s\","
             "\"agentId\":\"%s\",\"events\":%s}"),
        *BatchId, *ActiveSessionId, *ActiveAgentId, *EventsJson);

    FString Path = FString::Printf(
        TEXT("/sessions/%s/batch"), *ActiveSessionId);

    FAIArenaApiClient::PostRaw(Path, Body,
        FOnApiResponse::CreateLambda(
            [this, Batch](bool bOK, TSharedPtr<FJsonObject>)
            {
                if (bOK)
                {
                    RetryCount = 0;
                    UE_LOG(LogTemp, Verbose,
                           TEXT("[TelemetryCollector] Flushed %d events."),
                           Batch.Num());
                }
                else if (RetryCount < MaxRetries)
                {
                    RetryCount++;
                    UE_LOG(LogTemp, Warning,
                           TEXT("[TelemetryCollector] Flush failed (retry %d/%d). "
                                "Re-queuing %d events."),
                           RetryCount, MaxRetries, Batch.Num());
                    // Re-queue failed events at the front
                    EventBuffer.Insert(Batch, 0);
                }
                else
                {
                    UE_LOG(LogTemp, Error,
                           TEXT("[TelemetryCollector] Dropping %d events after %d retries."),
                           Batch.Num(), MaxRetries);
                    RetryCount = 0;
                }
            }));
}

File: unity/AIArenaSDK/Runtime/Telemetry/TelemetryCollector.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;
using AIArena.SDK.Core;

namespace AIArena.SDK.Telemetry
{
    /// <summary>
    /// Singleton telemetry collector. Records game events and batches them for submission.
    /// </summary>
    public class TelemetryCollector : MonoBehaviour
    {
        private static TelemetryCollector _instance;
        public static TelemetryCollector Instance
        {
            get
            {
                if (_instance == null)
                {
                    var go = new GameObject("[TelemetryCollector]");
                    _instance = go.AddComponent<TelemetryCollector>();
                    DontDestroyOnLoad(go);
                }
                return _instance;
            }
        }

        private AIArenaConfig _config;
        private readonly List<TelemetryEvent> _buffer = new List<TelemetryEvent>();
        private string _sessionId;
        private string _agentId;
        private int _sequenceNumber;
        private float _lastFlushTime;

        public void Configure(AIArenaConfig config)
        {
            _config = config;
            _lastFlushTime = Time.time;
        }

        public void StartSession(string sessionId, string agentId)
        {
            _sessionId = sessionId;
            _agentId = agentId;
            _sequenceNumber = 0;
            _buffer.Clear();
            Debug.Log($"[TelemetryCollector] Session started: {sessionId}");
        }

        /// <summary>Record a combat action event.</summary>
        public void RecordCombatAction(string actionType, string targetId, Vector3 position, bool success, float damageDealt, float latencyMs)
        {
            Enqueue("COMBAT_ACTION", new {
                actionType, targetId, position = ToDict(position),
                success, damageDealt, latencyMs,
            });
        }

        /// <summary>Record a position update event.</summary>
        public void RecordPositionUpdate(Vector3 position, Vector3 velocity, float rotation)
        {
            Enqueue("POSITION_UPDATE", new {
                position = ToDict(position),
                velocity = ToDict(velocity),
                rotation,
            });
        }

        /// <summary>Record an ability use event.</summary>
        public void RecordAbilityUse(string abilityId, string abilityName, string targetId, float cooldownMs)
        {
            Enqueue("ABILITY_USE", new { abilityId, abilityName, targetId, cooldownMs });
        }

        /// <summary>Record a health change event.</summary>
        public void RecordHealthChange(float previousHp, float currentHp, float maxHp, string changeReason, string sourceId = null)
        {
            Enqueue("HEALTH_CHANGE", new { previousHp, currentHp, maxHp, changeReason, sourceId });
        }

        /// <summary>Record a kill event.</summary>
        public void RecordKill(string victimId, string weaponUsed, Vector3 position, int killStreak)
        {
            Enqueue("KILL", new { victimId, weaponUsed, position = ToDict(position), killStreak });
        }

        private void Enqueue(string eventType, object payload)
        {
            if (!_config?.EnableTelemetry ?? false) return;

            var evt = new TelemetryEvent
            {
                EventId = Guid.NewGuid().ToString(),
                SessionId = _sessionId,
                AgentId = _agentId,
                EventType = eventType,
                Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                SequenceNumber = _sequenceNumber++,
                Payload = payload,
            };

            _buffer.Add(evt);

            if (_buffer.Count >= _config.TelemetryBatchSize)
            {
                _ = Flush();
            }
        }

        private void Update()
        {
            if (_config != null && Time.time - _lastFlushTime >= _config.TelemetryFlushIntervalSeconds)
            {
                _ = Flush();
            }
        }

        public async Task Flush()
        {
            if (_buffer.Count == 0 || _sessionId == null) return;

            var batch = new List<TelemetryEvent>(_buffer);
            _buffer.Clear();
            _lastFlushTime = Time.time;

            try
            {
                var batchData = new TelemetryBatch
                {
                    BatchId = $"{_sessionId}-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}",
                    SessionId = _sessionId,
                    AgentId = _agentId,
                    Events = batch.ToArray(),
                    SubmittedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                };

                await ApiClient.Post<object>($"/sessions/{_sessionId}/batch", batchData, _config);
                Debug.Log($"[TelemetryCollector] Flushed {batch.Count} events.");
            }
            catch (Exception ex)
            {
                Debug.LogError($"[TelemetryCollector] Flush failed: {ex.Message}");
                // Re-queue failed events
                _buffer.InsertRange(0, batch);
            }
        }

        public async Task FlushAll()
        {
            await Flush();
        }

        private static object ToDict(Vector3 v) => new { x = v.x, y = v.y, z = v.z };
    }

    [Serializable]
    public class TelemetryEvent
    {
        public string EventId;
        public string SessionId;
        public string AgentId;
        public string EventType;
        public long Timestamp;
        public int SequenceNumber;
        public object Payload;
    }

    [Serializable]
    public class TelemetryBatch
    {
        public string BatchId;
        public string SessionId;
        public string AgentId;
        public TelemetryEvent[] Events;
        public long SubmittedAt;
    }
}
